File: tests/perf/04_dml_perf.py
"""Track C performance -- DML PLR.

Times sp.dml(model='plr') with linear-regression nuisance learners
(closed-form, no learner Monte Carlo) at N in {1000, 5000, 10000}.
"""

from __future__ import annotations

import doubleml as dml_ref
import numpy as np
import pandas as pd
from _common import TimingResult, time_repeat, write_results
from sklearn.linear_model import LinearRegression

import statspai as sp

N_LIST = [1_000, 5_000, 10_000]
N_REPS = 3


def make_data(n: int, p: int = 5, seed: int = 42) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    X = rng.normal(size=(n, p))
    d = (X[:, 0] + 0.5 * X[:, 1] + rng.normal(size=n)) > 0
    d = d.astype(int)
    y = 0.5 * d + X.sum(axis=1) * 0.1 + rng.normal(size=n)
    cols = {f"x{i+1}": X[:, i] for i in range(p)}
    cols["d"] = d
    cols["y"] = y
    return pd.DataFrame(cols)


def main() -> None:
    rows: list[TimingResult] = []
    reference_rows: list[TimingResult] = []
    for n in N_LIST:
        df = make_data(n=n)

        def run_sp() -> None:
            sp.dml(
                data=df,
                y="y",
                d="d",
                X=[f"x{i+1}" for i in range(5)],
                model="plr",
                model_y=LinearRegression(),
                model_d=LinearRegression(),
            )

        med, iqr, mn, mx, peak = time_repeat(run_sp, n_reps=N_REPS, warmup=1)
        rows.append(
            TimingResult(
                estimator="04_dml",
                side="py",
                n=n,
                n_reps=N_REPS,
                median_time_s=med,
                iqr_time_s=iqr,
                min_time_s=mn,
                max_time_s=mx,
                peak_mem_mb=peak,
                extra={
                    "model": "plr",
                    "ml_g": "LinearRegression",
                    "ml_m": "LinearRegression",
                    "n_folds": 5,
                },
            )
        )

        def run_doubleml_python() -> None:
            # Fix fold assignment as well as the DGP. Object construction is
            # included on both sides because the public workflow is timed.
            np.random.seed(42)
            dml_data = dml_ref.DoubleMLData(
                df,
                y_col="y",
                d_cols="d",
                x_cols=[f"x{i+1}" for i in range(5)],
            )
            model = dml_ref.DoubleMLPLR(
                dml_data,
                ml_l=LinearRegression(),
                ml_m=LinearRegression(),
                n_folds=5,
            )
            model.fit()

        rmed, riqr, rmn, rmx, rpeak = time_repeat(
            run_doubleml_python, n_reps=N_REPS, warmup=1
        )
        reference_rows.append(
            TimingResult(
                estimator="04_dml",
                side="doubleml_py",
                n=n,
                n_reps=N_REPS,
                median_time_s=rmed,
                iqr_time_s=riqr,
                min_time_s=rmn,
                max_time_s=rmx,
                peak_mem_mb=rpeak,
                extra={
                    "model": "plr",
                    "reference": "doubleml-for-py",
                    "ml_l": "LinearRegression",
                    "ml_m": "LinearRegression",
                    "n_folds": 5,
                    "fold_seed": 42,
                },
            )
        )
        print(f"  N={n:>6}  StatsPAI={med:.3f}s  doubleml-for-py={rmed:.3f}s")

    write_results("04_dml", "py", rows)
    write_results("04_dml", "doubleml_py", reference_rows)


if __name__ == "__main__":
    main()

File: tests/perf/04_dml_perf.R
# Track C performance -- DML PLR (R side).
# Times DoubleML::DoubleMLPLR with mlr3::regr.lm at N in {1k, 5k, 10k}.

.script_dir <- (function() {
  args <- commandArgs(trailingOnly = FALSE)
  m <- grep("^--file=", args, value = TRUE)
  if (length(m) > 0) dirname(normalizePath(sub("^--file=", "", m[1])))
  else getwd()
})()
RESULTS_DIR <- file.path(.script_dir, "results")
dir.create(RESULTS_DIR, showWarnings = FALSE, recursive = TRUE)

suppressPackageStartupMessages({
  library(DoubleML)
  library(mlr3)
  library(mlr3learners)
  library(data.table)
  library(jsonlite)
  lgr::get_logger("mlr3")$set_threshold("warn")
})

N_LIST <- c(1000L, 5000L, 10000L)
N_REPS <- 3L

make_data <- function(n, p = 5L, seed = 42L) {
  set.seed(seed)
  X <- matrix(rnorm(n * p), nrow = n, ncol = p)
  d <- as.integer((X[, 1] + 0.5 * X[, 2] + rnorm(n)) > 0)
  y <- 0.5 * d + rowSums(X) * 0.1 + rnorm(n)
  out <- as.data.table(X)
  setnames(out, paste0("x", seq_len(p)))
  out[, d := d]
  out[, y := y]
  out
}

time_one <- function(fn, n_reps, warmup = 1L) {
  for (i in seq_len(warmup)) fn()
  t <- numeric(n_reps)
  for (i in seq_len(n_reps)) {
    gc()
    t0 <- Sys.time(); fn()
    t[i] <- as.numeric(Sys.time() - t0, units = "secs")
  }
  list(median = median(t), iqr = IQR(t), min = min(t), max = max(t))
}

rows <- list()
for (n in N_LIST) {
  dt <- make_data(n)
  fn <- function() {
    dml_data <- DoubleML::DoubleMLData$new(
      data = dt, y_col = "y", d_cols = "d",
      x_cols = paste0("x", 1:5)
    )
    dml_obj <- DoubleML::DoubleMLPLR$new(
      data = dml_data,
      ml_l = mlr3::lrn("regr.lm"),
      ml_m = mlr3::lrn("regr.lm"),
      n_folds = 5L
    )
    dml_obj$fit()
  }
  res <- time_one(fn, n_reps = N_REPS)
  rows[[length(rows) + 1L]] <- list(
    estimator = jsonlite::unbox("04_dml"),
    side = jsonlite::unbox("R"),
    n = jsonlite::unbox(as.integer(n)),
    n_reps = jsonlite::unbox(N_REPS),
    median_time_s = jsonlite::unbox(res$median),
    iqr_time_s = jsonlite::unbox(res$iqr),
    min_time_s = jsonlite::unbox(res$min),
    max_time_s = jsonlite::unbox(res$max),
    peak_mem_mb = jsonlite::unbox(NA_real_),
    extra = list(model = jsonlite::unbox("plr"),
                 ml_g = jsonlite::unbox("regr.lm"),
                 n_folds = jsonlite::unbox(5L))
  )
  cat(sprintf("  N=%6d  median=%.3fs\n", n, res$median))
}

payload <- list(estimator = jsonlite::unbox("04_dml"),
                side = jsonlite::unbox("R"), rows = rows,
                hardware = list(reference_package = jsonlite::unbox(paste("DoubleML", as.character(utils::packageVersion("DoubleML")))),
                R_version = jsonlite::unbox(R.version$version.string)),
                extra = list())
writeLines(
  jsonlite::toJSON(payload, pretty = TRUE, na = "null", null = "null", digits = NA),
  file.path(RESULTS_DIR, "04_dml_R.json")
)
message("OK -- wrote 04_dml_R.json")
